crowbar_framework/lib/crowbar/repository.rb

```diff
@@ -0,0 +1,131 @@
+#
+# Copyright 2015, SUSE LINUX GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module Crowbar
+  class Repository
+    attr_reader :platform, :id, :registry, :available, :md5, :url
+
+    class << self
+      #
+      # class methods
+      #
+      def load_registry
+        @@config ||= YAML.load_file(File.join(Rails.root, "config", "repos.yml"))
+      end
+
+      def where(options = {})
+        platform = options.fetch :platform, nil
+        repo = options.fetch :repo, nil
+        check_all_repos.select do |r|
+          if platform
+            if repo
+              r.platform == platform && r.name == repo
+            else
+              r.platform == platform
+            end
+          else
+            r.name == repo
+          end
+        end
+      end
+
+      def all_platforms
+        @@config.keys
+      end
+
+      def check_all_repos
+        repochecks = []
+        all_platforms.each do |platform|
+          repositories(platform).each do |repo|
+            check = self.new(platform, repo)
+            repochecks << check
+          end
+        end
+        repochecks
+      end
+
+      def admin_platform
+        NodeObject.admin_node.target_platform
+      end
+
+      def web_port
+        Proposal.where(barclamp: "provisioner").first.properties["web_port"]
+      end
+
+      def repositories(platform)
+        @@config[platform]["repos"].keys
+      end
+    end
+
+    #
+    # constructor
+    #
+    def initialize(platform, repo)
+      @platform = platform
+      @id = repo
+      @registry = Repository.load_registry[@platform]["repos"][@id]
+      @available = valid_repo_dir?
+      @md5 = valid_key_file?
+      @url = repo_url
```

Comment:
OK, so here we have a method `repo_url` to calculate the URL value, which is only used in this constructor, and the result is cached in `@url` and exposed via the `#url` accessor.  So both methods return the same value, and it is not clear to a consumer of this class which one should be used.  That's not ideal and it's the kind of thing which would have been more obvious if you'd followed my suggestion to do TDD and write the tests first, because then the first question would have been "what's the nicest interface for this class?"  (You may have heard this approach sometimes described as "write the code you wish you had" - it's an "outside-in" approach where first you write tests for an interface you wish you had, and then you implement the interface and watch the tests go from red to green.)

Anyway, what's the best way forward?  I think it depends on whether it makes sense to cache the calculated URL value or not:

- If you want to cache it, then the `#url` accessor should be the only way to retrieve the value, and `#repo_url` is an implementation detail which should become a private method (and is probably worth renaming to `#calculate_url` or similar).
- OTOH, if you don't want to cache it, then simply drop the existing `#url` accessor and `@url` instance variable, and rename `#repo_url` to `#url`.  (The `repo_` prefix is redundant because it's being called on a `Repository` object.)

Personally I don't see any benefit from caching it, and it makes the code a bit more complicated, so I'd prefer the second option.  But you should decide which you prefer.